components/payments/cmd/connectors/internal/connectors/atlar/task_fetch_payments.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/formancehq/payments/cmd/connectors/internal/ingestion"
+	"github.com/formancehq/payments/cmd/connectors/internal/metrics"
+	"github.com/formancehq/payments/cmd/connectors/internal/task"
+	"github.com/formancehq/payments/internal/models"
+	"github.com/formancehq/stack/libs/go-libs/logging"
+	"github.com/formancehq/stack/libs/go-libs/pointer"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+var (
+	paymentsAttrs = metric.WithAttributes(append(connectorAttrs, attribute.String(metrics.ObjectAttributeKey, "payments"))...)
+)
+
+func FetchPaymentsTask(config Config, account string, client *client.Client) task.Task {
+	return func(
+		ctx context.Context,
+		logger logging.Logger,
+		connectorID models.ConnectorID,
+		resolver task.StateResolver,
+		scheduler task.Scheduler,
+		ingester ingestion.Ingester,
+		metricsRegistry metrics.MetricsRegistry,
+	) error {
+		now := time.Now()
+		defer func() {
+			metricsRegistry.ConnectorObjectsLatency().Record(ctx, time.Since(now).Milliseconds(), paymentsAttrs)
+		}()
+
+		// Pagination works by cursor token.
+		params := transactions.GetV1TransactionsParams{
+			Context: ctx,
+			Limit:   pointer.For(int64(config.ApiConfig.PageSize)),
+		}
+		for token := ""; ; {
+			params.Token = &token
+			pagedTransactions, err := client.Transactions.GetV1Transactions(&params)
+			if err != nil {
+				metricsRegistry.ConnectorObjectsErrors().Add(ctx, 1, paymentsAttrs)
+				return err
+			}
+
+			token = pagedTransactions.Payload.NextToken
+
+			if err := ingestPaymentsBatch(ctx, connectorID, ingester, metricsRegistry, pagedTransactions); err != nil {
+				return err
+			}
+			metricsRegistry.ConnectorObjects().Add(ctx, int64(len(pagedTransactions.Payload.Items)), paymentsAttrs)
+
+			if token == "" {
+				break
+			}
+		}
+
+		return nil
+	}
+}
+
+func ingestPaymentsBatch(
+	ctx context.Context,
+	connectorID models.ConnectorID,
+	ingester ingestion.Ingester,
+	metricsRegistry metrics.MetricsRegistry,
+	pagedTransactions *transactions.GetV1TransactionsOK,
+) error {
+	batch := ingestion.PaymentBatch{}
+
+	for _, item := range pagedTransactions.Payload.Items {
+		raw, err := json.Marshal(item)
+		if err != nil {
+			return err
+		}
+
+		paymentType := determinePaymentType(item)
+
+		itemAmount := item.Amount
+		precision := supportedCurrenciesWithDecimal[*itemAmount.Currency]
+
+		var amount big.Float
+		_, ok := amount.SetString(*itemAmount.StringValue)
+		if !ok {
+			return fmt.Errorf("failed to parse amount %s", *itemAmount.StringValue)
+		}
+
+		var amountInt big.Int
+		amount.Mul(&amount, big.NewFloat(math.Pow(10, float64(precision)))).Int(&amountInt)
+
+		createdAt, err := ParseAtlarTimestamp(item.Created)
+		if err != nil {
+			return err
+		}
+
+		paymentId := models.PaymentID{
+			PaymentReference: models.PaymentReference{
+				Reference: item.ID,
+				Type:      paymentType,
+			},
+			ConnectorID: connectorID,
+		}
+
+		batchElement := ingestion.PaymentBatchElement{
```

Comment:
Since transactions can change over time, we'd want to overwrite any previously ingested payments. Currently this does not happen, instead trying to overwrite an ingested payment results in the following:

<img width="1588" alt="Screenshot 2023-11-28 at 16 54 24" src="https://github.com/formancehq/stack/assets/91877506/9e412a9f-e209-4375-ad6d-5f47ff1dfea5">

@paul-nicolas do you have any idea on how to forcibly overwrite any payments already in the database?

In the future we might want to only fetch changes since the last poling. We'd have to save the time of last polling somewhere in order to make the connector resilient against interruptions of service. This is an optimization for later. Would we need to optimize this before merging, though?

Reply:
Actually, you have another error on balances from what I can see on your screenshot ! Let me check this error and go back to you. 

Otherwise, payments can be overwritten overtime (in order to update statuses etc...)

When ingesting payments, you can add a state in order be able to fetch only the last changes since the last polling. We are using it on Stripe actually.

Reply:
We fixed the balance problem, I think the next optimizations can be in other PRs to not have a big one at the end !